### PR TITLE
V2 filesystem and guest module upgrade

### PR DESCRIPTION
**Pull request by bxy4543**
V2 filesystem and guest module upgrade
#873

### CODE REVIEW

File: pkg/filesystem/filesystem.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+		var err error
+		err = utils.Retry(10, time.Second, func() error {
+			err = mount.NewMountDriver().Unmount(mountdir)
+			if err != nil {
+				return err
+			}
+			return os.RemoveAll(mountdir)
+		})
+		if err != nil {
+			logger.Warn("failed to unmount dir %s,err: %v", mountdir, err)
```
Comment (allencloud) [inline pkg/filesystem/filesystem.go:69]: Shall we `return err` here?

File: pkg/filesystem/filesystem.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	}*/
+	var wg sync.WaitGroup
+	var flag bool
+	var mutex sync.Mutex
+	src := common.DefaultMountCloudImageDir(cluster.Name)
+	// TODO scp sdk has change file mod bug
+	initCmd := fmt.Sprintf(RemoteChmod, target)
+	for _, ip := range ipList {
+		wg.Add(1)
+		go func(ip string) {
```
Comment (allencloud) [inline pkg/filesystem/filesystem.go:158]: please distinguish the ip parameter from ip parameter at L156.

File: pkg/filesystem/filesystem.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+			defer wg.Done()
+			ssh, err := ssh.GetHostSSHClient(ip, cluster)
+			if err != nil {
+				logger.Error("get host ssh client failed %v", err)
+				mutex.Lock()
+				flag = true
```
Comment (allencloud) [inline pkg/filesystem/filesystem.go:164]: shall we just return when flag is true? I found that the code continue running in the goroutine. Is it reasonable?

File: pkg/filesystem/filesystem.go (hunk 1 above)
Comment (fanux) [reply pkg/filesystem/filesystem.go:69]: +1

File: pkg/filesystem/filesystem_test.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	}
+	tests := []struct {
+		name    string
+		arg     args
+		wantErr bool
+	}{
+		{
+			name: "test mount",
+			arg: args{
+				cluster: &v2.Cluster{
```
Comment (fanux) [inline pkg/filesystem/filesystem_test.go:39]:
```
cluster := &v2.Cluster{.....}

args{
   cluster:cluster,
}
```

Too many indentation.

File: pkg/filesystem/filesystem.go (hunk 3 above)
Comment (fanux) [reply pkg/filesystem/filesystem.go:162]: Using channel to handle goroutine error

File: pkg/filesystem/filesystem.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+				mutex.Unlock()
+				return
+			}
+			err = CopyFiles(ssh, ip == config.IP, ip, src, target)
+			if err != nil {
+				logger.Error("copy rootfs failed %v", err)
+				mutex.Lock()
```
Comment (fanux) [inline pkg/filesystem/filesystem.go:169]: Using error channel

**Review by fanux — APPROVED** [review]

**Pull request merged**